File: 19-Iterator/src/go/forward_iterator.go
package main

// ForwardIterator walks the playlist front to back. Its own cursor lives
// here, so multiple forward iterators can run independently.
type ForwardIterator struct {
	songs []Song
	index int
}

func (f *ForwardIterator) HasNext() bool {
	return f.index < len(f.songs)
}

func (f *ForwardIterator) Next() Song {
	song := f.songs[f.index]
	f.index++
	return song
}

File: 19-Iterator/src/rust/forward_iterator.rs
use crate::song::Song;

/// Concrete Iterator — walks the playlist front to back. Implementing the
/// standard `Iterator` trait means `for`, `take`, `map`, `filter`, etc.
/// all work for free. `next() -> Option` fuses `hasNext` + `next` into one.
pub struct ForwardIterator<'a> {
    songs: &'a [Song],
    index: usize,
}

impl<'a> ForwardIterator<'a> {
    pub fn new(songs: &'a [Song]) -> Self {
        ForwardIterator { songs, index: 0 }
    }
}

impl<'a> Iterator for ForwardIterator<'a> {
    type Item = &'a Song;

    fn next(&mut self) -> Option<&'a Song> {
        if self.index < self.songs.len() {
            let song = &self.songs[self.index];
            self.index += 1;
            Some(song)
        } else {
            None
        }
    }
}
